### PR TITLE
4.x - Cookie defaults

### PR DESCRIPTION
**Pull request by ADmad**
Currently it was a bit tedious having to specify common non default options/attributes when creating cookies.

For e.g. if your app is under a subfolder of webserver's document root a sensible default is to have the path for all cookies set to `/subfolder` instead of `/` (that's what `CookieComponent` did). But in 4.x you would have to specify the subfolder path each time when creating a new cookie.

So I have added a new static method `Cookie::setDefaults()` to avoid having to repeatedly pass common options.

### CODE REVIEW

File: src/Http/Cookie/Cookie.php [hunk 1]
```diff
@@ -160,6 +176,17 @@ public function __construct(
...
 
+    /**
+     * Set default options for the cookies.
+     *
+     * @param array $options Default options.
+     * @return void
+     */
+    public static function setDefaults(array $options): void
+    {
+        static::$defaults = $options + static::$defaults;
```
Comment (othercorey) [inline src/Http/Cookie/Cookie.php]: If you set defaults twice with a sparse array, do you expect it to reset the other fields or merge with previously set defaults?

File: src/Http/Cookie/Cookie.php [hunk 2]
```diff
@@ -160,6 +176,17 @@ public function __construct(
         $this->expiresAt = $expiresAt;
     }
 
+    /**
+     * Set default options for the cookies.
```
Comment (othercorey) [inline src/Http/Cookie/Cookie.php]: Is there a docblock explaining the allowed fields?

File: tests/TestCase/Http/Cookie/CookieTest.php [hunk 3]
```diff
@@ -510,4 +511,17 @@ public function testCreateFromHeaderString()
         // https://tools.ietf.org/html/draft-west-first-party-cookies-07#section-4.1
         $this->assertNull($result->getSameSite());
     }
+
+    public function testDefaults()
+    {
+        Cookie::setDefaults(['path' => '/cakephp', 'expires' => time()]);
```
Comment (markstory) [inline tests/TestCase/Http/Cookie/CookieTest.php]: We should have a test for invalid same site and expires values.

Comment (ADmad) [reply tests/TestCase/Http/Cookie/CookieTest.php]: Convenience of array means losing benefit of type hinting :sob:. I need array shapes :smile:.

Comment (ADmad) [reply tests/TestCase/Http/Cookie/CookieTest.php]: I added checks and tests :)

File: src/Http/Cookie/Cookie.php [hunk 4]
```diff
@@ -206,13 +233,50 @@ public static function create(string $name, $value, array $options = [])
...
+        }
+
+        if ($expires instanceof DateTimeInterface) {
+            /** @psalm-suppress UndefinedInterfaceMethod */
+            return $expires->setTimezone(new DateTimeZone('GMT'));
+        }
+
+        if (!is_string($expires) && !is_int($expires)) {
+            throw new InvalidArgumentException(sprintf(
+                'Invalid type `%s` for expires.',
```
Comment (markstory) [inline src/Http/Cookie/Cookie.php]:
```suggestion
                'Invalid type `%s` for expires. Expected an string, integer or DateTime object.',
```

**Review by markstory — APPROVED** [review]

**Pull request merged**